File: websocket.js
const W3CWebSocket = require('websocket').w3cwebsocket;
const WebSocketAsPromised = require('websocket-as-promised');


function externallyResolvablePromise () {

    let resolve, reject, bogusTimeout

    const p = new Promise ((resolve_, reject_) => {
        resolve = resolve_
        reject = reject_
        bogusTimeout = setTimeout (function () {}, 1000000000) // prevents unwanted process termination when no more event loop tasks left
    })

    p.resolved = false

    p.resolve = function () {
        p.resolved = true
        clearTimeout (bogusTimeout)
        resolve.apply (this, arguments)
    }

    p.reject = function () {
        p.resolved = true
        clearTimeout (bogusTimeout)
        reject.apply (this, arguments)
    }

    return p
}

class UnifiedWebsocket {
    constructor (uri) {
        this._wsp = new WebSocketAsPromised (uri, {
            createWebSocket: url => new W3CWebSocket(url)
        });
        this.ws = this._wsp.ws
        this.queue = []
        this.callbacks = []
        // Translates push style into pull style
        this.addCallback((message) => { if (this.queue.length > 0) { this.queue.shift ().resolve (message)} })
    }

    async isOpen () {
        return this._wsp.isOpened || this._wsp.closing
    }

    async send (data) {
        await this.connect ()
        this._wsp.send (data)
    }

    async recv () {
        await this.connect ()
        let promise = new externallyResolvablePromise ()
        this.queue.push (promise)
        return promise
    }

    async open () {
        if (!this.isOpen ()) {
            return await this._wsp.open()
        }
    }

    addCallback (callback) {
        this.callbacks.push (callback)
        this._wsp.onMessage.addListener (callback)
    }

    async connect () {
        await this._wsp.open()
    }

    async close () {
        await this._wsp.close()
    }
}

const myWs = new UnifiedWebsocket ('wss://stream.binance.com:9443/ws/ethbtc@depth');
myWs.addCallback((msg) => { console.log ("Inside callback: " + msg) })


;(async function foo () {
    await myWs.open ()
    console.log(await myWs.recv ())
    console.log (await myWs.recv ())
    console.log (await myWs.recv ())
    await myWs.close ()
}) ()

File: websocket.py
import websockets
import asyncio


class UnifiedWebsocket:
    def __init__(self, uri, protocol=None):
        self.uri = uri
        self.protocol = protocol
        self._connection = None
        self.callbacks = []
        self.ws = websockets.connect(self.uri, subprotocols=self.protocol)
        asyncio.ensure_future(self.callbackLoop())
        self.lastFuture = None
        self.loop = asyncio.get_event_loop()  # should be able to pass it
        self.lock = asyncio.Lock()

    async def open(self):
        async with self.lock:
            if self._connection is not None:
                return self._connection
            self._connection = await self.ws.__aenter__()
            return self._connection

    async def send(self, data):
        await self.open()
        return await self._connection.send(data)

    async def recv(self):
        await self.open()
        return await self.lastFuture

    def isOpen(self):
        return self._connection is not None and self._connection.open

    def addCallback(self, callback):
        self.callbacks.append(callback)

    async def close(self):
        async with self.lock:
            await self._connection.close()
            await self.ws.__aexit__(None, None, None)

    async def callbackLoop(self):
        """Translates pull style into push style"""
        await self.open()
        while True:
            async with self.lock:
                if self.isOpen():
                    self.lastFuture = loop.create_task(self._connection.recv())
                    msg = await self.lastFuture
                    for callback in self.callbacks:
                        callback(msg)
                else:
                    break


def onMessage(msg):
    print('inside callback: ' + msg)


myWs = UnifiedWebsocket('wss://stream.binance.com:9443/ws/ethbtc@kline_1d')
myWs.addCallback(onMessage)


async def test ():
    await myWs.open()
    print(await myWs.recv())
    print(await myWs.recv())
    print(await myWs.recv())
    await myWs.close()


loop = asyncio.get_event_loop()
asyncio.ensure_future(test(), loop=loop)
loop.run_forever()